frmts/vrt/vrtexpression_muparser.cpp

```diff
@@ -0,0 +1,156 @@
+ *
+ * Project:  Virtual GDAL Datasets
+ * Purpose:  Implementation of GDALExpressionEvaluator.
+ * Author:   Daniel Baston
+ *
+ ******************************************************************************
+ * Copyright (c) 2024, ISciences LLC
+ *
+ * SPDX-License-Identifier: MIT
+ ****************************************************************************/
+
+#include "vrtexpression.h"
+#include "cpl_string.h"
+
+#include <map>
+#include <muParser.h>
+
+namespace gdal
+{
+
+class MuParserExpression::Impl
+{
+  public:
+    explicit Impl(std::string_view osExpression)
+        : m_osExpression(std::string(osExpression)), m_oVectors{}, m_oParser{},
+          m_adfResults{1}, m_bIsCompiled{false}
+    {
+    }
+
+    void Register(std::string_view osVariable, double *pdfValue)
+    {
+        m_oParser.DefineVar(std::string(osVariable), pdfValue);
+    }
+
+    CPLErr Compile()
+    {
+        try
+        {
+            CPLString tmpExpression(m_osExpression);
+
+            for (const auto &[osVec, osElems] : m_oVectors)
+            {
+                tmpExpression.replaceAll(osVec, osElems);
+            }
+
+            m_oParser.SetExpr(tmpExpression);
+        }
+        catch (const mu::Parser::exception_type &e)
+        {
+            CPLError(CE_Failure, CPLE_AppDefined, "%s", e.GetMsg().c_str());
+            return CE_Failure;
+        }
+        catch (const std::exception &e)
+        {
+            CPLError(CE_Failure, CPLE_AppDefined, "%s", e.what());
+            return CE_Failure;
+        }
+
+        return CE_None;
+    }
+
+    CPLErr Evaluate()
+    {
+        if (!m_bIsCompiled)
+        {
+            if (auto eErr = Compile(); eErr != CE_None)
+            {
+                return eErr;
+            }
+
+            m_bIsCompiled = true;
+        }
+
+        try
+        {
+            m_adfResults[0] = m_oParser.Eval();
+        }
+        catch (const mu::Parser::exception_type &e)
+        {
+            CPLError(CE_Failure, CPLE_AppDefined, "%s", e.GetMsg().c_str());
+            return CE_Failure;
+        }
+        catch (const std::exception &e)
+        {
+            CPLError(CE_Failure, CPLE_AppDefined, "%s", e.what());
+            return CE_Failure;
+        }
+
+        return CE_None;
+    }
+
+    CPLString m_osExpression;
+    std::map<CPLString, CPLString> m_oVectors;
+    mu::Parser m_oParser;
+    std::vector<double> m_adfResults;
+    bool m_bIsCompiled;
+};
+
+MuParserExpression::MuParserExpression(std::string_view osExpression)
+    : m_pImpl{std::make_unique<Impl>(osExpression)}
+
+{
+}
+
+MuParserExpression::~MuParserExpression()
+{
+}
+
+CPLErr MuParserExpression::Compile()
+{
+    return m_pImpl->Compile();
+}
+
+void MuParserExpression::RegisterVariable(std::string_view osVariable,
+                                          double *pdfValue)
+{
+    m_pImpl->Register(osVariable, pdfValue);
+}
+
+void MuParserExpression::RegisterVector(std::string_view osVariable,
```

Comment:
not that it likely matters much performance wise, but given that you have to std::string(osVariable) at least twice in this method (including one in a loop), just taking a const std::string& as a parameter could be slightly better here

Reply:
Excessive copies avoided in 8d977e8b00dfac7cec952d094fa4c7bf27bf3c01 without changing interface